TumbleBitSetup/PoupardStern.cs

```diff
@@ -0,0 +1,263 @@
+﻿using Org.BouncyCastle.Math;
+using Org.BouncyCastle.Security;
+using Org.BouncyCastle.Utilities;
+using System;
+
+
+namespace TumbleBitSetup
+{
+    public class PoupardStern
+    {
+        /// <summary>
+        /// Proving Algorithm specified in page 12 (3.2.1) of the setup
+        /// </summary> 
+        /// <param name="p">P in the secret key</param>
+        /// <param name="q">Q in the secret key</param>
+        /// <param name="e">Public Exponent in the public key</param>
+        /// <param name="ps">The public string from the setup</param>
+        /// <param name="k">Security parameter as specified in the setup.</param>
+        /// <returns>List of x values and y</returns>
+        public static Tuple<BigInteger[], BigInteger> Proving(BigInteger p, BigInteger q, BigInteger e, byte[] psBytes, int k = 128)
+        {
+            BigInteger y;
+            BigInteger Two = BigInteger.Two;
+
+            // Generate a keyPair from p, q and e
+            var keyPair = new RsaKey(p, q, e);
+            var pubKey = new RsaPubKey(keyPair._pubKey);
+            var secKey = keyPair._privKey;
+
+            BigInteger Modulus = pubKey._pubKey.Modulus;
+            int ModulusBitLength = Modulus.BitLength;
+
+            // Calculating 2^{ |N| - 1}
+            BigInteger lowerLimit = Two.Pow(ModulusBitLength - 1);
+            
+            // Calculating phi
+            BigInteger pSub1 = p.Subtract(BigInteger.One);
+            BigInteger qSub1 = q.Subtract(BigInteger.One);
+            BigInteger phi = pSub1.Multiply(qSub1);
+
+            // Check that N>2^{|N|-1}
+            if (!(Modulus.CompareTo(lowerLimit) > 0))
+                throw new ArgumentOutOfRangeException("Bad RSA modulus N");
+
+            // (N-phi)*2^k << N
+            if (!(Modulus.Subtract(phi).Multiply(Two.Pow(k)).CompareTo(Modulus) < 0))
```

Comment:
Yes, the test passes this check. Please note that to avoid handling any small values (2^{-10}), I made the following assumption
[(N-phi(N)) * 2^k / N] < [2^{-10}] is the same as this [N/(N-phi(N)) * 2^k] > 2^10 (I flipped both sides and switched it to "greater than")
I think the above holds true, please correct me if I'm wrong.
3744b747cb73d64bc92dda52d3b77cbb61338d1a